File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        n = len(matrix)
        m = len(matrix[0])

        rs, re = 0, n-1
        cs, ce = 0, m-1
        ans = []

        while len(ans)<n*m:
            # l to r
            for i in range(cs, ce+1):
                ans.append(matrix[rs][i])
            rs+=1

            if len(ans) == n*m:
                break

            # top to down
            for i in range(rs, re+1):
                ans.append(matrix[i][ce])
            ce-=1

            if len(ans) == n*m:
                break

            # right to left
            for i in range(ce, cs-1, -1):
                ans.append(matrix[re][i])
            re-=1

            if len(ans) == n*m:
                break

            for i in range(re, rs-1, -1):
                ans.append(matrix[i][cs])
            cs+=1

        return ans

File: 0054-spiral-matrix/0054-spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& m) {
        int left=0, top=0;
        int bottom=m.size()-1, right = m[0].size()-1;
        vector<int>a;

        while(left <= right && top <= bottom){
            
            //left to right
            for( int j=left; j<=right; j++){
                a.push_back(m[top][j]);
            }
            top++;

            //top to bottom
            for( int i=top; i<=bottom; i++){
                a.push_back(m[i][right]);
            }
            right--;


            //right to left
            if(top<=bottom){
                for( int j=right; j>=left; j--){
                    a.push_back(m[bottom][j]);
                }
                bottom--;
            }

            //bottom to top
            if(left <= right){
                for( int i=bottom; i>=top; i--){
                a.push_back(m[i][left]);
                }
                left++;
            }
        }
        return a;
    }
};